samcli/lib/utils/osutils.py

```diff
@@ -62,3 +67,71 @@ def stderr():
         Byte stream of stderr
     """
     return sys.stderr.buffer
+
+
+def remove(path):
+    if path:
+        try:
+            os.remove(path)
+        except OSError:
+            pass
+
+
+@contextlib.contextmanager
+def tempfile_platform_independent():
+    # NOTE(TheSriram): Setting delete=False is specific to windows.
+    # https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile
+    _tempfile = tempfile.NamedTemporaryFile(delete=False)
+    try:
+        yield _tempfile
+    finally:
+        _tempfile.close()
+        remove(_tempfile.name)
+
+
+# NOTE: Py3.8 or higher has a ``dir_exist_ok=True`` parameter to provide this functionality.
+#       This method can be removed if we stop supporting Py37
+def copytree(source, destination, ignore=None):
+    """
+    Similar to shutil.copytree except that it removes the limitation that the destination directory should
+    be present.
+    :type source: str
+    :param source:
+        Path to the source folder to copy
+    :type destination: str
+    :param destination:
+        Path to destination folder
+    :type ignore: function
+    :param ignore:
+        A function that returns a set of file names to ignore, given a list of available file names. Similar to the
+        ``ignore`` property of ``shutils.copytree`` method
+    """
+
+    if not os.path.exists(destination):
+        os.makedirs(destination)
+
+        try:
+            # Let's try to copy the directory metadata from source to destination
+            shutil.copystat(source, destination)
+        except OSError as ex:
```

Comment:
this seems broad to be just windows errors, can we get more granular?

Reply:
this is not even my code. it's copy pasted ;)